packages/bp/src/core/messaging/messaging-service.ts

```diff
@@ -138,16 +152,37 @@ export class MessagingService {
     }
 
     const payloadAbsoluteUrl = this.convertToAbsoluteUrls(event.payload)
+
+    const collecting = event.incomingEventId && this.collectingCache.get(event.incomingEventId)
     const message = await this.clientsByBotId[event.botId].messages.create(
       event.threadId!,
       undefined,
-      payloadAbsoluteUrl
+      payloadAbsoluteUrl,
+      collecting ? { incomingId: collecting } : undefined
     )
     event.messageId = message.id
 
     return next(undefined, true, false)
   }
 
+  public informProcessingDone(event: IO.IncomingEvent) {
+    if (this.collectingCache.get(event.id!)) {
+      // We don't want the waiting for the queue to be empty to freeze other messages
+      // eslint-disable-next-line @typescript-eslint/no-floating-promises
+      this.sendProcessingDone(event)
+    }
+  }
+
+  private async sendProcessingDone(event: IO.IncomingEvent) {
+    try {
+      await this.eventEngine.waitOutgoingQueueEmpty(event)
+      await this.clientsByBotId[event.botId].messages.endTurn(event.messageId!)
+      this.collectingCache.del(event.id!)
```

Comment:
Shouldn't we delete the element from the cache no matter what? So maybe inside a finally block?

Reply:
It's not really critical for this code but makes sense why not